src/_pytest/terminal.py

```diff
@@ -588,7 +596,8 @@ def pytest_runtest_logstart(
             self.flush()
 
     def pytest_runtest_logreport(self, report: TestReport) -> None:
-        self._tests_ran = True
+        if report.when == "teardown":
+            self._all_tests_finished = True
```

Comment:
Note that `pytest_runtest_logreport` is not a good heuristic to set `_all_tests_finished` to True:

* This hook is called multiple times for each test
* Even for the teardown event, this might be called multiple times per test (for example pytest-repeat does this).

TBH I'm not a big fan on how this is turning out, IME every time state is being represented by a bunch of "control flags" like in here means a maintenance problem later... but I do not have a concrete suggestion either, the fact that you had to use those control flags suggest that the entire code needs a proper rethinking, giving it just grown organically over the years...

What do you think?

Reply:
I will be honest- I agree. 
After hours of thinking this was the simplest solution I found. 
So.. it fixes the bug completly- but at what cost? Also- the current bug started from the exact reason you just mentioned- we are update currentfspath control flag, which is kind of an anti pattern- or trying to escape from design. In the current implementation we are sometimes updating currentfsflag to -2 (what does it mean? What is this magical number?).

I can go to the journey of entire refactor of the code to fix it, but Im new here, so I tried to keep things simple. 
What do you think?

Reply:
@Oreldm I'm relieved we are on the same page. :sweat_smile: 

Not only this is critical section of the code (so any bugs here will reach almost any pytest run out there), I know there are plugins and IDEs that access those internals today (not to their fault, we do not have a public API and sometimes this is the best they can do), so any refactoring here is really risky.

TBH, I would abort this and leave the bug in place for now, it is nothing critical and the solution to refactor everything is not worth the risk I think.

Reply:
Sorry you have wasted so much time here, but in the end I think it is better to leave this bug alone for now.

Reply:
Well... Unlucky start for me, but maybe next time :) Aborted

Reply:
Yeah please do not get discouraged to contribute... definitely bad luck. 😁 

Cheers and thanks for being so understanding.